src/math/sat.rs

```diff
@@ -1,36 +1,252 @@
+            .collect();
+        let mut dedup_axes = Vec::new();
+        for ax1 in all_axes {
+            let is_dupe = dedup_axes.iter().any(|ax2: &Vector3<S>| {
+                let d1 = (ax1 - ax2).magnitude2();
+                let d2 = (ax1 + ax2).magnitude2();
+                d1.min(d2) < S::default_epsilon()
+            });
+            if !is_dupe {
+                dedup_axes.push(ax1);
+            }
         }
-        // Project corners of the box onto that axis
-        let mut box_min_proj: S = Float::max_value();
-        let mut box_max_proj: S = Float::min_value();
-        for corner in corners.iter() {
-            let corner_proj = corner.dot(*sep_axis);
-            box_min_proj = box_min_proj.min(corner_proj);
-            box_max_proj = box_max_proj.max(corner_proj);
+        CachedAxesIntersector {
+            axes: dedup_axes,
+            corners: self.corners,
         }
-        if box_min_proj > cube_max_proj || box_max_proj < cube_min_proj {
-            return false;
+    }
+
+    /// Perform an intersection test. The resulting Relation expresses how the other object is spatially
+    /// related to the self object – e.g. if `Relation::In` is returned, the other object is completely
+    /// inside the self object.
+    pub fn intersect(&self, other: &Self) -> Relation {
+        sat(
+            self.separating_axes_iter(&other.edges, &other.face_normals),
+            &self.corners,
+            &other.corners,
+        )
+    }
+}
+
+/// Stores pre-computed separating axes for intersection tests.
+pub struct CachedAxesIntersector<S: BaseFloat> {
+    pub axes: Vec<Vector3<S>>,
+    pub corners: [Point3<S>; 8],
+}
+
+impl<S: BaseFloat + Bounded> CachedAxesIntersector<S> {
+    /// Perform an intersection test using the cached axes and the specified corner points. The resulting
+    /// Relation expresses how the other object is spatially related to the self object – e.g. if
+    /// `Relation::In` is returned, the other object is completely inside the self object.
+    pub fn intersect(&self, corners: &[Point3<S>]) -> Relation {
+        sat(self.axes.iter().cloned(), &self.corners, corners)
+    }
+}
+
+/// See https://www.gamedev.net/forums/topic/694911-separating-axis-theorem-3d-polygons/ for more detail
+/// Return `Relation::In` if B is contained in A
+pub fn sat<S, I>(separating_axes: I, corners_a: &[Point3<S>], corners_b: &[Point3<S>]) -> Relation
+where
+    S: BaseFloat + Bounded,
+    I: IntoIterator<Item = Vector3<S>>,
+{
+    let mut rel = Relation::In;
+    for sep_axis in separating_axes {
+        // Project corners of A onto that axis
+        let mut a_min_proj: S = Bounded::max_value();
+        let mut a_max_proj: S = Bounded::min_value();
+        for corner in corners_a {
+            let corner_proj = corner.to_vec().dot(sep_axis);
+            a_min_proj = a_min_proj.min(corner_proj);
+            a_max_proj = a_max_proj.max(corner_proj);
+        }
```

Comment:
How about moving this into a helper function or closure taking the corners and separation axis, so it can be reused below?

Reply:
Ok

Reply:
Actually, I don't think that buys us readability, one updates a_min/max_proj, the other b_min/max_proj

Reply:
Ah, I see, they could be initialized inside and returned

Reply:
Done.